specs/agents/metrics.md

```diff
@@ -50,11 +50,109 @@ Agents should not send the `system.process.cgroup.memory.mem.limit.bytes` metric
+
+### Agent Health Metrics
+
+Agents SHOULD expose metrics which allow the analysis of the internal state of the agent, especially including the [reporting/transport mechanism](transport.md).
+
+Many of these metrics are `COUNTER`s: a `COUNTER` is a monotonically increasing metric.
+It is periodically reported by sending the difference of the counter-value now vs its value at the time of the previous report.
+In order to save elasticsearch disk space, zero-increases are not sent.
+
+Example:
+| **Time** | **Counter** | **Reported Metric Value** |
+|----------|-------------|---------------------------|
+| 0s       | 7           | 7                         |
+| 30s      | 10          | 3                         |
+| 1m       | 10          | 0 (not sent)              |
+| 1m30s    | 12          | 2                         |
+
+
+#### Event Count Metrics
+
+Agents send their data to the Intake API of the APM-server in the form of a stream of events (e.g. transactions, spans, metricsets...).
+
+The following metrics counting these events SHOULD be exposed:
+ - `agent.events.total`: `COUNTER` of events either submitted to the internal reporting queue or dropped upon submit attempt
+ - `agent.events.dropped.queue`: `COUNTER` of events where the attempt to add it to the reporting queue failed (e.g. due to a full queue)
+ - `agent.events.dropped.error`: `COUNTER` of events dropped after they previously have been successfully added to the queue.
+
+If an agent supports these metrics, the metrics MUST be captured for all events which would reach the APM-Server in an ideal, failure-free world.
+Therefore an event MUST be taken into account for these metrics if an attempt is made to submit it to the internal reporting queue, regardless of the success.
+Events dropped before this attempt (e.g. due to [sampling](tracing-sampling.md) or [transaction_max_spans](handling-huge-traces/tracing-spans-limit.md)) are NOT counted.
+
+The value of `agent.events.dropped.error` is an upper bound for the actual number of dropped events after entering the queue.
+For example, when an request to the Intake API fails, all events within this request MUST be considered as failed.
+In reality, it is possible that e.g. half of the data was actually successfully received and forwarded by the APM server.
+
+All these metrics MUST have the label `eventType`. The possible values for `eventType` are the lowercase names of the events according to the [spec](https://github.com/elastic/apm-server/tree/main/docs/spec/v2) (e.g. `transaction`, `metricset`) or the value `other`.
+
+Agents SHOULD attempt to assign the appropriate label value based on the counted event. If this would impose significant implementation overhead, the value `other` MUST be used instead.
```

Comment:
Do you have an example of where the overhead of determining the event type would cause overhead?

Reply:
With overhead I meant overhead in development time.

For example, metricsets / metadata are currently not distinguishable in the Java Agent at the point in time where they are added to the queue, this would need a bit of refactoring. That's why I though of leaving the `other` label as an escape hatch here in order to allow agents to implement the metrics initially without causing too much headache.

Reply:
The `JSON_WRITER` event is exclusively used for metrics. So you can safely assume `JSON_WRITER` is equal to `metricset`. Additionally, you could rename the event type to something like `METRICSET_JSON_WRITER` to make it a bit more clear.